### PR TITLE
AArch64: Add fixed format support for ACL primtives

### PR DESCRIPTION
**Pull request by jondea**
# Description

This PR makes the Compute Library for Arm® (ACL) primitives which use GEMM (inner product, matmul, conv) make use of fixed format kernels. By this we mean that the weights are reordered to be blocked/interleaved in oneDNN rather than ACL, so the kernel used needs to a have a known and fixed format.
This has multiple positives:
- It fits better with the oneDNN philosophy of separating the primitive and reorder
- Weights can be variable, because the reorder is performed at the oneDNN/framework level
- And therefore, most importantly: primitives can be properly cached in TensorFlow without checking the content of the weights (which is very hacky)

This has a negligible effect on performance overall, because the kernels are similar and the reorders are just happening in a different place.

The broad idea is:
- Ask ACL what `WeightsFormat` it wants the weights in, via the `has_opt_impl` function
- Convert the resulting `WeightsFormat` into a memory descriptor
- Get oneDNN to reorder by modifying the weights memory descriptor
- Set the strides on the weights TensorInfo to reflect the reordered layout

Things to note:
- This patch causes the examples `cpu-cnn-inference-f32-cpp` and `cpu-primitives-inner-product-cpp` to fail because inner product now implicitly collapses dimensions. This behaviour is supported by benchdnn, but not these two examples. Would it make sense to modify `reorder_primitive_desc_create` and `consistent_with` to allow collapsing dimensions?
- We have also removed wino because it is not yet compatible with this approach, and it is also failing for some unrelated reasons. Due to the existing heuristics we do not see it used in any models currently, but we may return to this later with a working implementation and more effective heuristics.
- This patch causes some NCHW fastmath convolutions to no longer be supported, we are working on a fix, but these are rarely used.

# Checklist

## General

- [x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit? (see comment in body)
- [X] Have you formatted the code using clang-format?

### CODE REVIEW

**Comment by jondea** [comment]
Any feedback would be appreciated on this. Also, would it still be possible for a backport to v3.1?

File: src/common/reorder.cpp [hunk 1]
```diff
@@ -70,16 +71,23 @@ status_t reorder_primitive_desc_create(std::shared_ptr<primitive_desc_t> &pd,
                     s_ek != d_ek, utils::one_of(engine_kind::cpu, s_ek, d_ek));
     if (!args_ok) return invalid_arguments;
 
-    auto s_mdw = memory_desc_wrapper(*src_md);
     auto d_mdw = memory_desc_wrapper(*dst_md);
 
-    if (!s_mdw.consistent_with(d_mdw)) return invalid_arguments;
+    // If src and dst are not already consistent, try to reshape them
+    memory_desc_t reshaped_src_md = *src_md;
+    auto reshaped_s_mdw = memory_desc_wrapper(&reshaped_src_md);
```
Comment (igorsafo) [inline src/common/reorder.cpp:78]: This change applies to the whole library and changes API behavior.
If you want to have such feature we will need an RFC to discuss available options.

As an alternative option you could uncollapse weights md internally before returning to a user (or keep both collapsed and uncollapsed versions for internal use and API calls). This way reorders will work as expected and the change will be isolated to this specific inner product implementation.

File: src/cpu/aarch64/acl_convolution_utils.cpp [hunk 2]
```diff
@@ -247,76 +281,38 @@ status_t init_conf_indirect_gemm(acl_conv_conf_t &acp, memory_desc_t &src_md,
...
                                 acp.dilation_info,
                                 acp.act_info,
                                 acp.fast_math,
-                                1)));
-    // clang-format on
-
-    return status::success;
-}
-
-status_t init_conf_wino(acl_conv_conf_t &acp, memory_desc_t &src_md,
```
Comment (igorsafo) [inline src/cpu/aarch64/acl_convolution_utils.cpp:277]: Did you remove Winograd implementation completely? If so, please move it to a separate commit.

**Comment by igorsafo** [comment]
> Any feedback would be appreciated on this. Also, would it still be possible for a backport to v3.1?

There is not much time to get it into v3.1 (Production release March 23, 2023 (ww12.4'23)), we are at the end of a validation cycle.
For this to happen we need to make sure:
- all tests pass on your side
- the changes are isolated to Aarch64 implementation

File: src/common/reorder.cpp (hunk 1 above)
Comment (jondea) [reply src/common/reorder.cpp:78]: That's a fair point, I'll have a think and create an RFC if it still makes sense.

Would you be able to expand on what you mean by this

> As an alternative option you could uncollapse weights md internally in the implementation before returning to a user 

I think we need to pass the collapsed memory descriptor back to the user so that they can do the right reorder. The collapsed reorder is a different to the non-collapsed one. It's also beneficial, for example If we have a tensor with H=2, W=2 and IC=3 and we need to pad to 4, then
- 2x2x3 -pad-> 2x2x4 -collapse-> 16
- 2x2x3 -collapse-> 12 -pad-> 12

File: src/cpu/aarch64/acl_inner_product.hpp [hunk 3]
```diff
@@ -98,6 +101,7 @@ struct acl_inner_product_fwd_t : public primitive_t {
                             primitive_attr_t::skip_mask_t::post_ops, f32);
             const bool ok = is_fwd() && !has_zero_dim_memory()
                     && utils::one_of(true, is_fp16_ok, is_fp32_ok)
+                    && weights_md_.format_kind == format_kind::any
```
[the remark this answers is not in the document]
Comment (jondea) [reply src/cpu/aarch64/acl_inner_product.hpp:104]: Yes.

I thought that using the `format::any` is best practice, and we expected that most users do this already, particularly for inference. This is certainly the case for TensorFlow, where this will allow the reorders to be hoisted out. Do you know whether users tend to do this?

Am I right in thinking that to support specific formats, we would need to implement the reorder logic inside each primitive?

**Comment by jondea** [comment]
Understood, thank you for the feedback @igorsafo!

File: src/cpu/aarch64/acl_inner_product.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (igorsafo) [reply src/cpu/aarch64/acl_inner_product.hpp:104]: > Yes.
> 
> I thought that using the `format::any` is best practice, and we expected that most users do this already, particularly for inference. This is certainly the case for TensorFlow, where this will allow the reorders to be hoisted out. Do you know whether users tend to do this?


You are right, the best practice and a programming model tells user to create inner product with `any` so oneDNN chooses the best implementation. However sometimes it is not possible on user's side (when weights are constant and can't be reordered by a framework) and oneDNN programming model supports passing plain layouts to inner product -- it will not return unimplemented/invalid arguments. After your changes inner product created with plain weights layout will dispatch into gemm-based or reference inner product which is totally fine, but might result in worse performance comparing to previos state. 

> Am I right in thinking that to support specific formats, we would need to implement the reorder logic inside each primitive?

This is a possible solution, but there are implications:
- scratchpad is required because primitive is stateless so we can't cache changes inside
- reorder will be required each execution due to stateless primitive


Again, no changes are required on your side, this is completely fine from oneDNN programming model I just wanted to make sure this is your intention (disabling explicit weight formats in acl ip implementation). You can resolve this discussion.

**Comment by jondea** [comment]
I should say that all the tests pass now on AArch64

File: src/cpu/aarch64/acl_convolution_utils.cpp [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
-* Copyright 2020-2022 Arm Ltd. and affiliates
+* Copyright 2020-2023 Arm Ltd. and affiliates
```
Comment (igorsafo) [inline src/cpu/aarch64/acl_convolution_utils.cpp:2]: (random spot) Please update convolution documentation by removing Aarch64 from Winograd support https://github.com/oneapi-src/oneDNN/blob/master/doc/primitives/convolution.md

Comment (jondea) [reply src/cpu/aarch64/acl_convolution_utils.cpp:2]: Good spot! Done

**Review by igorsafo — APPROVED** [review]

**Review by mgouicem — APPROVED** [review]

**Comment by mgouicem** [comment]
THanks for the contribution!
